apps/prairielearn/src/pages/publicQuestionFileDownload/publicQuestionFileDownload.ts

```diff
@@ -0,0 +1,37 @@
+import { Router } from 'express';
+import asyncHandler from 'express-async-handler';
+
+import * as error from '@prairielearn/error';
+
+import { UserSchema } from '../../lib/db-types.js';
+import { selectCourseById } from '../../models/course.js';
+import { selectQuestionById } from '../../models/question.js';
+
+const router = Router({ mergeParams: true });
+
+async function setLocals(req, res) {
+  res.locals.user = UserSchema.parse(res.locals.authn_user);
+  res.locals.authz_data = { user: res.locals.user };
+  res.locals.course = await selectCourseById(req.params.course_id);
+  res.locals.question = await selectQuestionById(req.params.question_id);
+
+  if (
+    !res.locals.question.share_source_publicly ||
+    res.locals.course.id !== res.locals.question.course_id
+  ) {
+    throw new error.HttpStatusError(404, 'Not Found');
+  }
+  return;
+}
+
+router.get(
+  '/*',
+  asyncHandler(async (req, res) => {
+    await setLocals(req, res);
+    if (req.query.type) res.type(req.query.type.toString());
+    if (req.query.attachment) res.attachment(req.query.attachment.toString());
+    res.sendFile(req.params[0], { root: res.locals.course.path });
```

Comment:
This would allow _anyone_ to download _any_ file from a course with a publicly-shared question. You can reproduce this for yourself by manually constructing a URL, e.g. `http://localhost:3000/pl/public/course/.../question/.../file_download/infoCourse.json`. `infoCourse.json` and really most files shouldn't be downloadable via this route.

Reply:
I think that I've solved this by simply adding a call to `getPaths` to the download page. If not, please let me know and I can have another look at this.

Reply:
Also, I assume this is not a major concern because a user needs course viewer permissions anyway to download files, but the same trick is also possible for the existing instructor file download page. Should this be fixed as well (considering that it is just 1 extra line of code)?

Reply:
Yeah, I think we don't care about that because "view" permissions for a course are sufficient to download _any_ file in the course; there's no case where we'd really care if they downloaded some arbitrary file.

Reply:
It seems like this conversation should be marked 'resolved' now is that right?